benchmark/win.ipynb

```diff
@@ -0,0 +1,3654 @@
+     "status": "completed"
+    },
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# Split Data\n",
+    "train, test = train_test_split(verified_matches, test_size=0.33, random_state=SEED)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "eb83098e",
+   "metadata": {
+    "papermill": {
+     "duration": 0.02877,
+     "end_time": "2023-12-11T22:19:27.143623",
+     "exception": false,
+     "start_time": "2023-12-11T22:19:27.114853",
+     "status": "completed"
+    },
+    "tags": []
+   },
+   "source": [
+    "# Parse Data for OpenSkill Training Set\n",
+    "Let's now parse the training data for OpenSkill."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "id": "8d393d68",
+   "metadata": {
+    "execution": {
+     "iopub.execute_input": "2023-12-11T22:19:27.203573Z",
+     "iopub.status.busy": "2023-12-11T22:19:27.203144Z",
+     "iopub.status.idle": "2023-12-11T22:19:29.323119Z",
+     "shell.execute_reply": "2023-12-11T22:19:29.321489Z"
+    },
+    "papermill": {
+     "duration": 2.153397,
+     "end_time": "2023-12-11T22:19:29.326022",
+     "exception": false,
+     "start_time": "2023-12-11T22:19:27.172625",
+     "status": "completed"
+    },
+    "tags": []
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Loading Raw Data from Training Set into Memory\n",
+      "Parsing Training Data:\n"
+     ]
+    },
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "6a9b50d97c4c4be8a7d846424ea835a5",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "  0%|          | 0/3792 [00:00<?, ?it/s]"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Parsed Training Matches\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"Loading Raw Data from Training Set into Memory\")\n",
+    "\n",
+    "# Parse Training Data\n",
+    "print(\"Parsing Training Data:\")\n",
+    "\n",
+    "# Get Unique Matches for Training\n",
+    "train_size = len(train)\n",
+    "\n",
+    "# Create a Progress Bar\n",
+    "t = tqdm(total=train_size)\n",
+    "\n",
+    "os_process_time_start = time.time()\n",
+    "\n",
+    "for match in train:\n",
+    "    result = match.get(\"result\")\n",
+    "    won = True if result == \"WIN\" else False\n",
+    "\n",
+    "    teams: dict = match.get(\"teams\")\n",
+    "    blue_team: dict = teams.get(\"blue\")\n",
+    "    red_team: dict = teams.get(\"red\")\n",
+    "\n",
+    "    os_blue_players = {}\n",
+    "    os_red_players = {}\n",
+    "\n",
+    "    m = model\n",
+    "    r = m.rating\n",
+    "\n",
+    "    for player in blue_team:\n",
+    "        os_blue_players[player] = r()\n",
+    "\n",
+    "    for player in red_team:\n",
+    "        os_red_players[player] = r()\n",
```

Comment:
As far as I can see this creates a new rating object for each player even if they have been rated in a previous match? Is that intended? I would have though something like

```Python
for player in blue_team:
    os_blue_players[player] = openskill_players.setdefault(player, r())
for player in red_team:
    os_red_players[player] = openskill_players.setdefault(player, r())
```

would be more along the lines of what was wanted? (Same comment applies for updates to TrueSkill model below)